cmd/npmExecuteScripts.go

```diff
@@ -0,0 +1,166 @@
+	return u.fileUtils.FileExists(path)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) glob(pattern string) (matches []string, err error) {
+	return doublestar.Glob(pattern)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getwd() (dir string, err error) {
+	return os.Getwd()
+}
+
+func (u *npmExecuteScriptsUtilsBundle) chdir(dir string) error {
+	return os.Chdir(dir)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) usesMta() bool {
+	return u.projectStructure.UsesMta()
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getExecRunner() execRunner {
+	if u.execRunner == nil {
+		u.execRunner = &command.Command{}
+		u.execRunner.Stdout(log.Entry().Writer())
+		u.execRunner.Stderr(log.Entry().Writer())
+	}
+	return u.execRunner
+}
+
+func npmExecuteScripts(config npmExecuteScriptsOptions, telemetryData *telemetry.CustomData) {
+	utils := npmExecuteScriptsUtilsBundle{}
+
+	err := runNpmExecuteScripts(&utils, &config)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+func runNpmExecuteScripts(utils npmExecuteScriptsUtilsInterface, options *npmExecuteScriptsOptions) error {
+	execRunner := utils.getExecRunner()
+	log.Entry().Infof("NPM Registry configuration: defaultNpmRegistry %s, sapNpmRegistry %s",
+		options.DefaultNpmRegistry, options.SapNpmRegistry)
+	environment := []string{"npm_config_@sap:registry=" + options.SapNpmRegistry}
+	if options.DefaultNpmRegistry != "" {
+		environment = append(environment, "npm_config_registry="+options.DefaultNpmRegistry)
+	}
+	execRunner.SetEnv(environment)
+
+	packageJSONFiles, err := findPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+
+	for _, file := range packageJSONFiles {
+		dir := path.Dir(file)
+		err = utils.chdir(dir)
+		if err != nil {
+			return err
+		}
+		packageLockExists, err := utils.fileExists("package-lock.json")
+
+		if err != nil {
+			return err
+		}
+		yarnLockExists, err := utils.fileExists("yarn.lock")
+		if err != nil {
+			return err
+		}
+		if options.Install {
+			err = installDependencies(dir, packageLockExists, yarnLockExists, execRunner)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, v := range options.RunScripts {
+			log.Entry().WithField("WorkingDirectory", dir).Info("run-script " + v)
+			err = execRunner.RunExecutable("npm", "run-script", v, "--if-present")
+			if err != nil {
+				return err
+			}
+		}
+		err = utils.chdir(oldWorkingDirectory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func findPackageJSONFiles(utils npmExecuteScriptsUtilsInterface) ([]string, error) {
+	unfilteredListOfPackageJSONFiles, err := utils.glob("**/package.json")
+	if err != nil {
+		return nil, err
+	}
+
+	var packageJSONFiles []string
+
+	isMtaProject := utils.usesMta()
+	for _, file := range unfilteredListOfPackageJSONFiles {
+		if strings.Contains(file, "node_modules") {
+			continue
+		}
+		if isMtaProject && (strings.HasPrefix(file, "gen/") || strings.Contains(file, "/gen/")) {
```

Comment:
Why ignoring "gen/" in case of MTA. I think gen directories can also be generated  in non MTA cases. That is a cap feature not and mta feature.

What about letting the user define a list of buildDescribtor files which should be ignored, as we already have it at multiple places in piper

Reply:
I'm not sure if I understand the concept of a build descriptor. In go, I see that term in karma and a security scan, but it seems to be something different. Can you elaborate on how this should work?